### PR TITLE
Tensor revision

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

- Removed NNCFCollectorTensorProcessor (left implementation only for TF), replace to use tensor function instead
- Update ptq algorithms to using comon `Tensor` in not backend specific parts
- Add `expand_dims` to tensor.numeric 
- Change default value for `keepdims` for `numeric .quantile` function to be allying numpy signature

### Related tickets

113588

### CODE REVIEW

File: nncf/experimental/tensor/functions/numeric.py [hunk 1]
```diff
@@ -389,12 +389,16 @@ def mean(
 @tensor_guard
 def median(a: Tensor, axis: Optional[Union[int, Tuple[int, ...]]] = None, keepdims: bool = False) -> Tensor:
     """
-    Compute the arithmetic median along the specified axis.
+    <<<<<<< HEAD
```
Comment (alexsu52) [inline nncf/experimental/tensor/functions/numeric.py:392]: ?

File: nncf/experimental/tensor/functions/torch_numeric.py [hunk 2]
```diff
@@ -227,29 +227,30 @@ def _(
     a: torch.Tensor,
     q: Union[float, List[float]],
     axis: Optional[Union[int, Tuple[int]]] = None,
-    keepdims: Optional[bool] = None,
+    keepdims: Optional[bool] = False,
```
Comment (alexsu52) [inline nncf/experimental/tensor/functions/torch_numeric.py:230]: This change is not aligned with https://github.com/openvinotoolkit/nncf/blob/99b5b85aaa5ee0c200edfa3de1ef8ff27a640385/nncf/experimental/tensor/functions/numeric.py#L454

File: nncf/quantization/algorithms/bias_correction/algorithm.py [hunk 3]
```diff
@@ -352,7 +355,14 @@ def _create_feed_dicts(
                 # Since we do not use as inputs the layers from which the statistics are gathered,
                 # but those that follow them, we need to take this into account when creating feed dicts.
                 activation_name, _ = self._collected_stat_inputs_map[(input_node_name, input_port_id)]
-                feed_dict[input_tensor_name] = statistics_per_input[input_tensor_name][stat_id]
+
+                val = statistics_per_input[input_tensor_name][stat_id]
+                if isinstance(val, Tensor):
```
Comment (alexsu52) [inline nncf/quantization/algorithms/bias_correction/algorithm.py:360]: ?

File: nncf/onnx/statistics/collectors.py [hunk 4]
```diff
@@ -30,169 +26,12 @@
...
-        return ONNXNNCFTensor(a.tensor - b.tensor)
-
-    @staticmethod
-    def zero_elements(x: ONNXNNCFTensor) -> ONNXNNCFTensor:
-        np_tensor = x.tensor
-        eps = np.finfo(np_tensor.dtype).eps
-        return ONNXNNCFTensor(np.abs(np_tensor) < eps)
-
-
 class ONNXBasicReducer(TensorReducerBase):
```
Comment (daniil-lyakhov) [inline nncf/onnx/statistics/collectors.py:34]: Can we remove all `ONNX` and `PT` reducers and replace them by the base classes? And left reimplementation only for the `OV` as only the `OV` has inplace fns?

File: nncf/experimental/tensor/functions/torch_numeric.py (hunk 2 above)
Comment (AlexanderDokuchaev) [reply nncf/experimental/tensor/functions/torch_numeric.py]: Changed to `keepdims: bool = False,` according https://numpy.org/doc/stable/reference/generated/numpy.quantile.html

File: nncf/onnx/statistics/collectors.py (hunk 4 above)
Comment (AlexanderDokuchaev) [reply nncf/onnx/statistics/collectors.py:34]: Yes, it's possible but in other PR

Comment (daniil-lyakhov) [reply nncf/onnx/statistics/collectors.py:34]: But this classes become redundant in this pr, why should we split the work? Why we are refactoring only the one part of the reducers and keep the other? Do we really need a commit with redundant classes? 

We could just agree to keep redundant classes as it is, don't see the point to make a separate pr

**Review by alexsu52 — APPROVED** [review]
LGTM

**Review by daniil-lyakhov — APPROVED** [review]

File: nncf/quantization/fake_quantize.py [hunk 5]
```diff
@@ -240,8 +240,9 @@ def calculate_quantizer_parameters(
         False - the full range is used.
     :return: Parameters of the FakeQuantize layer.
     """
-    min_values = Tensor(statistics.min_values).astype(TensorDataType.float32)
-    max_values = Tensor(statistics.max_values).astype(TensorDataType.float32)
+    print(statistics.min_values)
```
Comment (kshpv) [inline nncf/quantization/fake_quantize.py:243]: debug

**Review by kshpv — CHANGES_REQUESTED** [review]

File: nncf/quantization/fake_quantize.py (hunk 5 above)
Comment (AlexanderDokuchaev) [reply nncf/quantization/fake_quantize.py]: removed

File: nncf/quantization/algorithms/bias_correction/algorithm.py [hunk 6]
```diff
@@ -175,17 +176,17 @@ def apply(
 
             channel_axis = node.metatype.output_channel_axis
             if current_bias.ndim > 1:
-                channel_axis = range(current_bias.ndim)[channel_axis]
-                axes = [i for i in range(current_bias.ndim) if i != channel_axis]
-                bias_shift = np.expand_dims(bias_shift, axes)
+                new_shape = [1] * current_bias.ndim
+                new_shape[channel_axis] = bias_shift.shape[0]
+                bias_shift = bias_shift.reshape(new_shape)
```
Comment (KodiaqQ) [inline nncf/quantization/algorithms/bias_correction/algorithm.py:181]: Is this equal change? Setting the channel axis for the new shape may not be the same as the shift's shape 0  element. Also, the expand_dims is not equal to reshape.

File: nncf/quantization/algorithms/bias_correction/algorithm.py [hunk 7]
```diff
@@ -175,17 +176,17 @@ def apply(
...
 
             updated_bias = current_bias + bias_shift
             magnitude = self._get_bias_shift_magnitude(current_bias, updated_bias)
 
             if magnitude < self.threshold:
                 nncf_logger.debug(f"{node_name} bias would be changed. Magnitude: {magnitude}")
                 bias_correction_command = self._backend_entity.create_bias_correction_command(
-                    node, updated_bias, nncf_graph
+                    node, updated_bias.data, nncf_graph
```
Comment (KodiaqQ) [inline nncf/quantization/algorithms/bias_correction/algorithm.py:189]: Why can't we pass the `Tensor` into transformation and unpack the value in the `ModelTransformer`?

File: nncf/quantization/algorithms/bias_correction/algorithm.py [hunk 8]
```diff
@@ -376,25 +377,23 @@ def _compute_bias_shift(
...
+            q_outputs.append(mean_per_channel(q_output, channel_axis))
         # Here we get the per-sample average, so the axis is 0.
-        q_output = np.mean(q_outputs, axis=0)
-        return output_fp - q_output
+        q_output = fns.mean(fns.stack(q_outputs), axis=0)
+        return fns.stack(output_fp) - q_output
 
     @staticmethod
-    def _get_bias_shift_magnitude(current_bias_value: np.ndarray, updated_bias_value: np.ndarray) -> float:
+    def _get_bias_shift_magnitude(current_bias_value: Tensor, updated_bias_value: Tensor) -> float:
```
Comment (KodiaqQ) [inline nncf/quantization/algorithms/bias_correction/algorithm.py:386]: Isn't the resulting value the `Tensor` now?

File: nncf/quantization/algorithms/min_max/onnx_backend.py [hunk 9]
```diff
@@ -154,16 +154,14 @@ def create_convert_insertion_command(
         raise nncf.InternalError("FakeConvert insertion not implemented in ONNX backend!")
 
     @staticmethod
-    def unify_statistics(
-        statistics: List[ONNXMinMaxTensorStatistic],
-    ) -> ONNXMinMaxTensorStatistic:
+    def unify_statistics(statistics: List[MinMaxTensorStatistic]) -> MinMaxTensorStatistic:
```
Comment (KodiaqQ) [inline nncf/quantization/algorithms/min_max/onnx_backend.py:157]: This method can be placed in the common algorithm part due to unification.

**Review by KodiaqQ — COMMENTED** [review]
General comment: the extension of the `experimental` code usage in the `common` part is not good at all. Are there any reasons to keep `Tensor` as the `experimental`, but spread it wider and wider?

File: nncf/experimental/common/tensor_statistics/collectors.py [hunk 10]
```diff
@@ -551,27 +538,28 @@ def __init__(
...
 
 
 class BatchMeanReducer(TensorReducerBase):
     def __init__(self, inplace: bool = False):
         super().__init__(None, inplace)
 
-    def _reduce_out_of_place(self, x: List[NNCFTensor]) -> List[NNCFTensor]:
-        return [self._tensor_processor.batch_mean(x[0])]
+    def _reduce_out_of_place(self, x: List[Tensor]) -> List[Tensor]:
+        # x[0] ???
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/experimental/common/tensor_statistics/collectors.py]: removed

File: nncf/quantization/algorithms/bias_correction/algorithm.py (hunk 6 above)
Comment (AlexanderDokuchaev) [reply nncf/quantization/algorithms/bias_correction/algorithm.py]: Revert changes and implement expand_dims to tensor.function. 
https://github.com/openvinotoolkit/nncf/pull/2689/commits/d03191d099fb403e3005aeb4a0ecd79ddbb45148#diff-48a31470e6f5691c217d11e1f28cfcf5dfa25ffdd047ca71353e0ec829b9485eR371-R388

File: nncf/quantization/algorithms/min_max/onnx_backend.py (hunk 9 above)
Comment (AlexanderDokuchaev) [reply nncf/quantization/algorithms/min_max/onnx_backend.py]: good suggestion, done

File: nncf/quantization/algorithms/bias_correction/algorithm.py (hunk 8 above)
Comment (AlexanderDokuchaev) [reply nncf/quantization/algorithms/bias_correction/algorithm.py]: fixed

File: nncf/quantization/algorithms/bias_correction/algorithm.py (hunk 7 above)
Comment (AlexanderDokuchaev) [reply nncf/quantization/algorithms/bias_correction/algorithm.py]: Change signature to `def create_bias_correction_command(node: NNCFNode, bias_value: Tensor) -> TransformationCommand:` and unwrap data in backend part.
But not think that `ModelTransformer` should operate with Tensor, it's backend specific part. at least there is no common commands.

My suggestion, that `algorithms.py` should works with tensors, `*_backend.py` unwrap data and call backend specific functions.

File: nncf/common/tensor_statistics/aggregator.py [hunk 11]
```diff
@@ -140,10 +141,10 @@ def _get_merged_statistic_points(
 
     @staticmethod
     @abstractmethod
-    def _process_outputs(outputs: Any) -> Dict[str, NNCFTensor]:
+    def _process_outputs(outputs: Any) -> Union[Dict[str, Tensor], Tensor]:
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/common/tensor_statistics/aggregator.py]: it's _prcoess_output was return tensor before, but not used to register statistics. 
Made some change to keep annotation
https://github.com/openvinotoolkit/nncf/pull/2689/commits/f74b3b9cc338edfe9a44371e6ec8db999d809e59#diff-6d22aadab242fbeb3d3a27095819407213185eff223221134f19c119fb50e8f5R71-R74

**Comment by AlexanderDokuchaev** [comment]
> General comment: the extension of the `experimental` code usage in the `common` part is not good at all. Are there any reasons to keep `Tensor` as the `experimental`, but spread it wider and wider?

@KodiaqQ planning do it after merge this PR, to moved Tensor from experimental directory and discuss about api, location and using.

**Review by andrey-churkin — APPROVED** [review]

**Review by daniil-lyakhov — APPROVED** [review]

**Pull request merged**